crypto/x509/v3_purp.c

```diff
@@ -629,17 +627,10 @@ int ossl_x509v3_cache_extensions(X509 *x)
      */
 #endif
     ERR_pop_to_mark();
-    if ((x->ex_flags & (EXFLAG_INVALID | EXFLAG_NO_FINGERPRINT)) == 0) {
-        CRYPTO_THREAD_unlock(x->lock);
-        return 1;
-    }
-    if ((x->ex_flags & EXFLAG_INVALID) != 0)
-        ERR_raise(ERR_LIB_X509, X509V3_R_INVALID_CERTIFICATE);
-    /* If computing sha1_hash failed the error queue already reflects this. */
-
- err:
-    x->ex_flags |= EXFLAG_SET; /* indicate that cert has been processed */
     CRYPTO_THREAD_unlock(x->lock);
+    if ((x->ex_flags & EXFLAG_INVALID) == 0)
+        return 1;
```

Comment:
Why is this check moved from lock?

Reply:
The lock is just in this function for consistently updating the cert structure.
Reading a single field, e.g., `ex_flags`, does not need such a protection. In  `crypto/x509/` you can find many such instances.

Reply:
I have some suspicions, but prefer not to change the behavior - e.g. to avoid thread races.

Reply:
So moved the check under the lock again.